runtime/compiler/control/CompileBeforeCheckpoint.cpp

```diff
@@ -0,0 +1,105 @@
+
+#include "j9cfg.h"
+
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+
+#include "j9protos.h"
+#include "j9nonbuilder.h"
+#include "control/CompilationController.hpp"
+#include "control/CompilationRuntime.hpp"
+#include "control/CompileBeforeCheckpoint.hpp"
+#include "control/OptimizationPlan.hpp"
+#include "env/Region.hpp"
+#include "env/VerboseLog.hpp"
+#include "env/VMJ9.h"
+
+TR::CompileBeforeCheckpoint::CompileBeforeCheckpoint(TR::Region &region, J9VMThread *vmThread, TR_J9VMBase *fej9, TR::CompilationInfo *compInfo) :
+   _region(region),
+   _vmThread(vmThread),
+   _fej9(fej9),
+   _compInfo(compInfo),
+   _methodList(_region),
+   _seenMethodsSet(std::less<TR_OpaqueMethodBlock*>(), _region)
+   {}
+
+void
+TR::CompileBeforeCheckpoint::addMethodToList(TR_OpaqueMethodBlock *method)
+   {
+   if (method && _seenMethodsSet.find(method) == _seenMethodsSet.end())
+      {
+      _methodList.push_back(method);
+      _seenMethodsSet.insert(method);
+      }
+   }
+
+void
+TR::CompileBeforeCheckpoint::queueMethodsForCompilationBeforeCheckpoint()
+   {
+   for (auto iter = _methodList.begin(); iter != _methodList.end(); iter++)
+      {
+      TR_OpaqueMethodBlock *method = *iter;
+      J9Method *j9method = reinterpret_cast<J9Method *>(method);
+
+      J9ROMMethod *romMethod = J9_ROM_METHOD_FROM_RAM_METHOD(j9method);
+      if (!_compInfo->isCompiled(j9method)
+          && !_J9ROMMETHOD_J9MODIFIER_IS_SET(romMethod, J9AccMethodFrameIteratorSkip)
+          && !_J9ROMMETHOD_J9MODIFIER_IS_SET(romMethod, J9AccAbstract))
+         {
+         if (TR::Options::getCmdLineOptions()->getVerboseOption(TR_VerboseCheckpointRestoreDetails))
+            {
+            TR_VerboseLog::CriticalSection cs;
+            TR_VerboseLog::write(TR_Vlog_CHECKPOINT_RESTORE, "Attempting to queue ");
+            _compInfo->printMethodNameToVlog(j9method);
+            TR_VerboseLog::writeLine(" (%p) for compilation", j9method);
+            }
+
+         TR_MethodEvent event;
+         event._eventType = TR_MethodEvent::CompilationBeforeCheckpoint;
+         event._j9method = j9method;
+         event._oldStartPC = 0;
+         event._vmThread = _vmThread;
+         event._classNeedingThunk = 0;
+         bool newPlanCreated;
+         TR_OptimizationPlan *plan = TR::CompilationController::getCompilationStrategy()->processEvent(&event, &newPlanCreated);
+         // the plan needs to be created only when async compilation is possible
+         // Otherwise the compilation will be triggered on next invocation
+         if (plan)
+            {
+            bool queued = false;
+            _fej9->startAsyncCompile(method, 0, &queued, plan);
+            if (!queued && newPlanCreated)
+               TR_OptimizationPlan::freeOptimizationPlan(plan);
+            }
+         }
+      }
+   }
+
+void
+TR::CompileBeforeCheckpoint::collectAndCompileMethodsBeforeCheckpoint()
+   {
+   collectMethodsForCompilationBeforeCheckpoint();
```

Comment:
While collecting the set of methods we must ensure that class unloading does not render any method invalid.

Reply:
Updated in https://github.com/eclipse-openj9/openj9/compare/d5850691d5686f23e30b6437904cf504f3a2a87e..fac845ac4f51b4ee18576fb1eb00fb9e002af68d